classes/Search.php

```diff
@@ -563,11 +565,43 @@ protected static function fillProductArray(&$product_array, $weight_array, $key,
 	public static function indexation($full = false, $id_product = false)
 	{
 		$db = Db::getInstance();
+		$all_product_shop = false;
 
 		if ($id_product)
 			$full = false;
 
-		if ($full)
+		if (Context::getContext()->shop->getContext() == Shop::CONTEXT_SHOP)
+		{
+			$id_shop = Context::getContext()->shop->getContextShopID();
+
+			$db->execute('DELETE si FROM `'._DB_PREFIX_.'search_index` si
+				INNER JOIN `'._DB_PREFIX_.'product` p ON (p.id_product = si.id_product)
+				'.Shop::addSqlAssociation('product', 'p').'
+				WHERE product_shop.`visibility` IN ("both", "search")
+				AND product_shop.`active` = 1
+				AND p.`id_product` IN (
+					SELECT * FROM (
+					SELECT p.`id_product` FROM `'._DB_PREFIX_.'product` p
+					'.Shop::addSqlAssociation('product', 'p').
+					' WHERE product_shop.`id_shop` = '.$id_shop.
+					') AS subquery
+				)');
```

Comment:
Actually after reading again the patch, I'm not sure what you are trying to do in this query.

If you want to delete the entries for product in more than one shop, you should try something like :

$db->execute('CREATE TEMPORARY TABLE `'._DB_PREFIX_.'multishop_products` SELECT id_product, COUNT(*) c FROM  `'._DB_PREFIX_.'product` p JOIN `'._DB_PREFIX_.'product_shop` ps ON (p.id_product=ps.id_product) GROUP BY p.id_product HAVING c>1;

$db->execute('DELETE si FROM `'._DB_PREFIX_.'search_index` si
                                INNER JOIN `'._DB_PREFIX_.'multishop_products` msp ON (si.id_product = msp.id_product)
				INNER JOIN `'._DB_PREFIX_.'product` p ON (p.id_product = si.id_product)
				'.Shop::addSqlAssociation('product', 'p').'
				WHERE product_shop.`visibility` IN ("both", "search")
				AND product_shop.`active` = 1');